File: week4/valid-parentheses.java
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> x = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                x.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (x.empty()) {
                    return false;
                }
                if (c == ')' && x.peek() != '(' ||
                    c == '}' && x.peek() != '{' ||
                    c == ']' && x.peek() != '[') {
                    return false;
                }
                x.pop();
            }
        }
        return x.empty();
    }
}

File: summer2021/unit4/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            ")": "(",
            "}": "{",
            "]": "["
        }
        openings = set(["(", "{", "["])
        stack = []
        for i in s:
            if i in openings:
                stack.append(i)
            elif i in pairs:
                if len(stack) == 0 or not pairs[i] == stack.pop():
                    return False
        return len(stack) == 0